plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/SizeBasedBlockingQ.java

```diff
@@ -0,0 +1,180 @@
+            executorService.submit(consumer);
+        }
+    }
+
+    public void produce(Item item) throws InterruptedException {
+        if (item == null || item.size <= 0) {
+            throw new IllegalStateException("Invalid item input to produce.");
+        }
+        final Lock lock = this.lock;
+        final AtomicLong currentSize = this.currentSize;
+        lock.lock();
+        try {
+            if (closed.get()) {
+                throw new AlreadyClosedException("Transfer queue is already closed.");
+            }
+            while (currentSize.get() + item.size >= capacity.getBytes()) {
+                notFull.await();
+            }
+            if (closed.get()) {
+                throw new AlreadyClosedException("Transfer queue is already closed.");
+            }
+            queue.put(item);
+            currentSize.addAndGet(item.size);
+            notEmpty.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public int getSize() {
+        return queue.size();
+    }
+
+    protected static class Consumer extends Thread {
+        private final LinkedBlockingQueue<Item> queue;
+        private final Lock lock;
+        private final Condition notFull;
+        private final Condition notEmpty;
+        private final AtomicLong currentSize;
+        private final AtomicBoolean closed;
+
+        public Consumer(
+            LinkedBlockingQueue<Item> queue,
+            AtomicLong currentSize,
+            Lock lock,
+            Condition notFull,
+            Condition notEmpty,
+            AtomicBoolean closed
+        ) {
+            this.queue = queue;
+            this.lock = lock;
+            this.notEmpty = notEmpty;
+            this.notFull = notFull;
+            this.currentSize = currentSize;
+            this.closed = closed;
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    consume();
+                } catch (AlreadyClosedException ex) {
+                    return;
+                } catch (Exception ex) {
+                    log.error("Failed to consume transfer event", ex);
+                }
+            }
+        }
+
+        private void consume() throws InterruptedException {
+            final Lock lock = this.lock;
+            final AtomicLong currentSize = this.currentSize;
+            lock.lock();
+            Item item;
+            try {
+                if (closed.get()) {
+                    throw new AlreadyClosedException("transfer queue closed");
+                }
+                while (currentSize.get() == 0) {
+                    notEmpty.await();
+                    if (closed.get()) {
+                        throw new AlreadyClosedException("transfer queue closed");
+                    }
+                }
+
+                item = queue.take();
+                currentSize.addAndGet(-item.size);
```

Comment:
Should be ok I think because if permits are not available consumer thread would wait. So, maximum items which can be put would still be queue size + (permits * avg size of an item). 
But coupling it with permits would add complexity to this queue plus by keeping threads waiting for permit we are being proactively ready for execution.